lib/rmg/RmgPath.cpp

```diff
@@ -0,0 +1,196 @@
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "RmgPath.h"
+#include <boost/heap/priority_queue.hpp> //A*
+
+using namespace rmg;
+
+const std::function<float(const int3 &, const int3 &)> Path::DEFAULT_MOVEMENT_FUNCTION =
+[](const int3 & src, const int3 & dst)
+{	
+	return 1.f;
+};
+
+//A* priority queue
+typedef std::pair<int3, float> TDistance;
+struct NodeComparer
+{
+	bool operator()(const TDistance & lhs, const TDistance & rhs) const
+	{
+		return (rhs.second < lhs.second);
+	}
+};
+boost::heap::priority_queue<TDistance, boost::heap::compare<NodeComparer>> createPriorityQueue()
+{
+	return boost::heap::priority_queue<TDistance, boost::heap::compare<NodeComparer>>();
+}
+
+Path::Path(const Area & area): dArea(&area)
+{
+}
+
+Path::Path(const Area & area, const int3 & src): dArea(&area)
+{
+	dPath.add(src);
+}
+
+Path::Path(const Path & path): dArea(path.dArea), dPath(path.dPath)
+{
+	
+}
+
+Path & Path::operator= (const Path & path)
+{
+	//do not modify area
+	dPath = path.dPath;
+	return *this;
+}
+
+bool Path::valid() const
+{
+	return !dPath.empty();
+}
+
+Path Path::invalid()
+{
+	return Path({});
+}
+
+Path Path::search(const Tileset & dst, bool straight, std::function<float(const int3 &, const int3 &)> moveCostFunction) const
+{
+	//A* algorithm taken from Wiki http://en.wikipedia.org/wiki/A*_search_algorithm
+	if(!dArea)
+		return Path::invalid();
+	
+	auto resultArea = *dArea + dst;
+	Path result(resultArea);
+	if(dst.empty())
+		return result;
+	
+	int3 src = rmg::Area(dst).nearest(dPath);
+	result.connect(src);
+	
+	Tileset closed;    // The set of nodes already evaluated.
+	auto open = createPriorityQueue(); // The set of tentative nodes to be evaluated, initially containing the start node
+	std::map<int3, int3> cameFrom;  // The map of navigated nodes.
+	std::map<int3, float> distances;
+	
+	cameFrom[src] = int3(-1, -1, -1); //first node points to finish condition
+	distances[src] = 0;
+	open.push(std::make_pair(src, 0.f));
+	// Cost from start along best known path.
+	
+	while(!open.empty())
+	{
+		auto node = open.top();
+		open.pop();
+		int3 currentNode = node.first;
+		
+		closed.insert(currentNode);
+		
+		if(dPath.contains(currentNode)) //we reached connection, stop
+		{
+			// Trace the path using the saved parent information and return path
+			int3 backTracking = currentNode;
+			while (cameFrom[backTracking].valid())
+			{
+				result.dPath.add(backTracking);
+				backTracking = cameFrom[backTracking];
+			}
+			return result;
+		}
+		else
+		{
+			auto foo = [&open, &closed, &cameFrom, &currentNode, &distances, &moveCostFunction, &result](const int3& pos) -> void
+			{
+				if(closed.count(pos))
+					return;
+				
+				if(!result.dArea->contains(pos))
+					return;
+				
+				float movementCost = moveCostFunction(currentNode, pos) + currentNode.dist2d(pos);
```

Comment:
currentNode.dist2d(pos) will return either 1 or 1.4 because these nodes are neighbours. Described A* algorithm has gScore and fScore where fScore is estimated distance from source to target through pos as far as I understood. But I do not see where is it implemented, am I missing something?
Btw does it search a new Path from current Path to dst or vise versa from dst to the current path?

Reply:
fScore is what called distance here and stored in distances map if I understood this.
As for path, it's starting search from dst to current Path and then backtracking result path from Path to dst. Overall it doesn't matter as it stored as std::set - just set of tiles connecting src and dot